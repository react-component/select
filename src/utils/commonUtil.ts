export function toArray<T>(value: T | T[]): T[] {
  if (Array.isArray(value)) {
    return value;
  }
  return value !== undefined ? [value] : [];
}

export const isClient =
  typeof window !== 'undefined' && window.document && window.document.documentElement;

/** Is client side and not jsdom */
export const isBrowserClient = process.env.NODE_ENV !== 'test' && isClient;
