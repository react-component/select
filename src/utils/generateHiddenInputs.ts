import type { DraftValueType } from '@/Select';

/**
 * @param inputNameAttributeVal - name value passed from <Select> component
 * @param value - the value or value(s) selected from the <Select> component
 * @param prefixCls
 *
 * @description
 * This function generates hidden `<input>` element(s) after a value is Selected from the <Select> component.
 * These hidden inputs serve as the value containers for the selected <Option>'s
 *
 * This is to achieve HTML5 `<form>` compatability & is an additive feature that is backwards compatiable with
 * existing ant's <Select> component. For morebackground context & history, see discussions below:
 *  - https://github.com/tailwindlabs/headlessui/pull/1214
 *  - https://github.com/react-component/select/issues/799
 */
export const generateHiddenInputs = (
  inputNameAttributeVal: string,
  value: string | string[] | DraftValueType[] | DraftValueType | any,
  prefixCls = 'rc-select',
) => {
  // We will append the hidden inputs below this element. The hidden inputs act as value container(s) for the selected option values.
  const rcSelectInputElement = document.querySelector(
    `.${prefixCls} input[class^="${prefixCls}"]`,
  ) as HTMLInputElement;
  if (!rcSelectInputElement) return;

  /**
   * Part 1:
   * Before creating any new hidden inputElements, first check if we have any existing inputs that were generated
   * from previously calling `onSelect`, `onDeselect`, `onChange` etc.
   */
  const hiddenInputCSSSelector = `.${prefixCls} input.hidden-input`;
  const hiddenInputElements = Array.from(document.querySelectorAll(hiddenInputCSSSelector));
  hiddenInputElements.forEach((hiddenInputElement) => hiddenInputElement.remove());

  // debugger;
  /**
   * Part 2:
   * Generate new hidden <input /> Elements and append them in the .rc-select container.
   */
  if (Array.isArray(value)) {
    value.forEach((val) => {
      const _val = val?.value ? val.value : val;
      const hiddenInputElement = document.createElement('input');
      hiddenInputElement.setAttribute('type', 'hidden');
      hiddenInputElement.setAttribute('name', inputNameAttributeVal);
      hiddenInputElement.setAttribute('value', _val);
      hiddenInputElement.setAttribute('class', 'hidden-input');
      rcSelectInputElement.insertAdjacentElement(`afterend`, hiddenInputElement);
    });
  } else {
    // Select value is single value
    const _val = value?.value ? value.value : value;
    const hiddenInputElement = document.createElement('input');
    hiddenInputElement.setAttribute('type', 'hidden');
    hiddenInputElement.setAttribute('name', inputNameAttributeVal);
    hiddenInputElement.setAttribute('value', _val);
    hiddenInputElement.setAttribute('class', 'hidden-input');
    rcSelectInputElement.insertAdjacentElement(`afterend`, hiddenInputElement);
  }
};
