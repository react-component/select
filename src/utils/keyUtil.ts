import KeyCode from 'rc-util/lib/KeyCode';

/** keyCode Judgment function */
export function isValidateOpenKey(currentKeyCode: number): boolean {
  // https://github.com/ant-design/ant-design/issues/45609
  return [
    KeyCode.ENTER
  ].includes(currentKeyCode);
}
