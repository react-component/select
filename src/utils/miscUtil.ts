export const handleClickAction = (fn, { isIE = false }) =>
  (isIE ? setTimeout(() => fn()) : fn());
