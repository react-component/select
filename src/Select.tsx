/**
 * To match accessibility requirement, we always provide an input in the component.
 * Other element will not set `tabIndex` to avoid `onBlur` sequence problem.
 * For focused select, we set `aria-live="polite"` to update the accessibility content.
 *
 * ref:
 * - keyboard: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/listbox_role#Keyboard_interactions
 *
 * New api:
 * - listHeight
 * - listItemHeight
 * - component
 *
 * Remove deprecated api:
 * - multiple
 * - tags
 * - combobox
 * - firstActiveValue
 * - dropdownMenuStyle
 * - openClassName (Not list in api)
 *
 * Update:
 * - `backfill` only support `combobox` mode
 * - `combobox` mode not support `labelInValue` since it's meaningless
 * - `getInputElement` only support `combobox` mode
 * - `onChange` return OptionData instead of ReactNode
 * - `filterOption` `onChange` `onSelect` accept OptionData instead of ReactNode
 * - `combobox` mode trigger `onChange` will get `undefined` if no `value` match in Option
 * - `combobox` mode not support `optionLabelProp`
 */

import * as React from 'react';
import { OptionsType as SelectOptionsType } from './interface';
import SelectOptionList from './OptionList';
import Option from './Option';
import OptGroup from './OptGroup';
import { convertChildrenToData as convertSelectChildrenToData } from './utils/legacyUtil';
import {
  getLabeledValue as getSelectLabeledValue,
  filterOptions as selectDefaultFilterOptions,
  isValueDisabled as isSelectValueDisabled,
  findValueOption as findSelectValueOption,
  flattenOptions,
  fillOptionsWithMissingValue,
} from './utils/valueUtil';
import generateSelector, { SelectProps, RefSelectProps } from './generate';
import { DefaultValueType, Key, SingleType } from './interface/generator';
import warningProps from './utils/warningPropsUtil';

const RefSelect = generateSelector<SelectOptionsType>({
  prefixCls: 'rc-select',
  components: {
    optionList: SelectOptionList,
  },
  convertChildrenToData: convertSelectChildrenToData,
  flattenOptions,
  getLabeledValue: getSelectLabeledValue,
  filterOptions: selectDefaultFilterOptions,
  isValueDisabled: isSelectValueDisabled,
  findValueOption: findSelectValueOption,
  warningProps,
  fillOptionsWithMissingValue,
});

export type ExportedSelectProps<
  ValueType extends DefaultValueType = DefaultValueType
> = SelectProps<SelectOptionsType, ValueType>;

/**
 * Typescript not support generic with function component,
 * we have to wrap an class component to handle this.
 */
class Select<
  VT extends Key | Key[] = Key,
  TKeyForOptions extends Key = Key
> extends React.Component<SelectProps<SelectOptionsType<TKeyForOptions, SingleType<VT>>, VT>> {
  static Option: typeof Option = Option;

  static OptGroup: typeof OptGroup = OptGroup;

  selectRef = React.createRef<RefSelectProps>();

  focus = () => {
    this.selectRef.current.focus();
  };

  blur = () => {
    this.selectRef.current.blur();
  };

  render() {
    return <RefSelect ref={this.selectRef} {...this.props} />;
  }
}

export default Select;
