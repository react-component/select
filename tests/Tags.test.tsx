/* eslint-disable import/no-named-as-default-member */
import { mount } from 'enzyme';
import KeyCode from 'rc-util/lib/KeyCode';
import classNames from 'classnames';
import * as React from 'react';
import Select, { OptGroup, Option } from '../src';
import allowClearTest from './shared/allowClearTest';
import blurTest from './shared/blurTest';
import dynamicChildrenTest from './shared/dynamicChildrenTest';
import focusTest from './shared/focusTest';
import hoverTest from './shared/hoverTest';
import inputFilterTest from './shared/inputFilterTest';
import openControlledTest from './shared/openControlledTest';
import removeSelectedTest from './shared/removeSelectedTest';
import renderTest from './shared/renderTest';
import throwOptionValue from './shared/throwOptionValue';
import { injectRunAllTimers, findSelection, expectOpen, toggleOpen } from './utils/common';

describe('Select.Tags', () => {
  injectRunAllTimers(jest);

  allowClearTest('tags', ['1128']);
  focusTest('tags', {});
  blurTest('tags');
  hoverTest('tags');
  renderTest('tags');
  removeSelectedTest('tags');
  throwOptionValue('tags');
  dynamicChildrenTest('tags', {});
  inputFilterTest('tags');
  openControlledTest('tags');

  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  it('allow user input tags', () => {
    const wrapper = mount(<Select mode="tags" />);

    wrapper
      .find('input')
      .simulate('change', { target: { value: 'foo' } })
      .simulate('keyDown', { which: KeyCode.ENTER });

    expect(findSelection(wrapper).text()).toBe('foo');
  });

  it('should call onChange on blur', () => {
    const onChange = jest.fn();
    const wrapper = mount(<Select mode="tags" onChange={onChange} />);

    wrapper
      .find('input')
      .simulate('change', { target: { value: 'foo' } })
      .simulate('blur');

    jest.runAllTimers();
    expect(findSelection(wrapper).text()).toBe('foo');
    expect(onChange).toHaveBeenCalledWith(['foo'], [{}]);
  });

  it('tokenize input', () => {
    const handleChange = jest.fn();
    const handleSelect = jest.fn();
    const option2 = <Option value="2">2</Option>;
    const wrapper = mount(
      <Select mode="tags" tokenSeparators={[',']} onChange={handleChange} onSelect={handleSelect}>
        <Option value="1">1</Option>
        {option2}
      </Select>,
    );

    wrapper.find('input').simulate('change', { target: { value: '2,3,4' } });

    expect(handleChange).toHaveBeenCalledWith(['2', '3', '4'], expect.anything());
    expect(handleSelect).toHaveBeenCalledTimes(3);
    expect(handleSelect).toHaveBeenLastCalledWith('4', expect.anything());
    expect(findSelection(wrapper).text()).toEqual('2');
    expect(findSelection(wrapper, 1).text()).toEqual('3');
    expect(findSelection(wrapper, 2).text()).toEqual('4');
    expect(wrapper.find('input').props().value).toBe('');
    expectOpen(wrapper, false);
  });

  it('should not separate words when compositing but trigger after composition end', () => {
    const handleChange = jest.fn();
    const handleSelect = jest.fn();
    const wrapper = mount(
      <Select mode="tags" tokenSeparators={[',']} onChange={handleChange} onSelect={handleSelect}>
        <Option value="1">1</Option>
        <Option value="2">2</Option>
      </Select>,
    );

    // composition start
    wrapper.find('input').simulate('compositionStart');
    wrapper.find('input').simulate('change', { target: { value: '2,3,4' } });
    expect(handleChange).not.toHaveBeenCalled();
    handleChange.mockReset();

    // composition end
    wrapper.find('input').simulate('compositionEnd');
    expect(handleChange).toHaveBeenCalledWith(['2', '3', '4'], expect.anything());

    expect(handleSelect).toHaveBeenCalledTimes(3);
    expect(handleSelect).toHaveBeenLastCalledWith('4', expect.anything());
    expect(findSelection(wrapper).text()).toEqual('2');
    expect(findSelection(wrapper, 1).text()).toEqual('3');
    expect(findSelection(wrapper, 2).text()).toEqual('4');
    expect(wrapper.find('input').props().value).toBe('');
    expectOpen(wrapper, false);
  });

  it('should work when menu is closed', () => {
    const handleChange = jest.fn();
    const handleSelect = jest.fn();
    const wrapper = mount(
      <Select
        mode="tags"
        tokenSeparators={[',']}
        onChange={handleChange}
        onSelect={handleSelect}
        open={false}
      >
        <Option value="1">1</Option>
        <Option value="2">2</Option>
      </Select>,
    );
    wrapper.find('input').simulate('compositionstart');
    wrapper.find('input').simulate('change', { target: { value: 'Star Kirby' } });
    wrapper.find('input').simulate('keydown', { which: KeyCode.ENTER });
    expect(handleChange).not.toHaveBeenCalled();
    handleChange.mockReset();
    wrapper.find('input').simulate('compositionend');
    wrapper.find('input').simulate('keydown', { which: KeyCode.ENTER });
    expect(handleChange).toHaveBeenCalledWith(['Star Kirby'], expect.anything());
    expect(handleSelect).toHaveBeenCalledTimes(1);
    expect(findSelection(wrapper).text()).toEqual('Star Kirby');
    expect(wrapper.find('input').props().value).toBe('');
    expectOpen(wrapper, false);
  });

  // Paste tests
  [
    {
      tokenSeparators: [' ', '\n'],
      clipboardText: '\n  light\n  bamboo\n  ',
      inputValue: '   light   bamboo   ',
    },
    {
      tokenSeparators: ['\r\n'],
      clipboardText: '\r\nlight\r\nbamboo\r\n',
      inputValue: ' light bamboo',
    },
    {
      tokenSeparators: [' ', '\r\n'],
      clipboardText: '\r\n light\r\n bamboo\r\n ',
      inputValue: '  light  bamboo  ',
    },
    {
      tokenSeparators: ['\n'],
      clipboardText: '\nlight\nbamboo\n',
      inputValue: ' light bamboo',
    },
  ].forEach(({ tokenSeparators, clipboardText, inputValue }) => {
    it(`paste content to split (${JSON.stringify(tokenSeparators)})`, () => {
      const onChange = jest.fn();
      const wrapper = mount(
        <Select mode="tags" tokenSeparators={tokenSeparators} onChange={onChange}>
          <Option value="1">1</Option>
        </Select>,
      );

      wrapper.find('input').simulate('paste', {
        clipboardData: {
          getData: () => clipboardText,
        },
      });
      wrapper.find('input').simulate('change', {
        target: { value: inputValue },
      });

      expect(onChange).toHaveBeenCalledWith(['light', 'bamboo'], expect.anything());
      expect(onChange).toHaveBeenCalledTimes(1);
    });
  });

  it('renders unlisted item in value', () => {
    const wrapper = mount(
      <Select mode="tags" value="3" open>
        <Option value="1">1</Option>
        <Option value="2">2</Option>
      </Select>,
    );

    expect(wrapper.find('List').props().data).toHaveLength(3);
  });

  it('dropdown keeps order', () => {
    const wrapper = mount(<Select mode="tags" open value={['aaaaa', 'aaa']} />);

    wrapper.find('input').simulate('change', { target: { value: 'aaa' } });
    expectOpen(wrapper);
    expect(wrapper.find('List').props().data).toEqual([
      expect.objectContaining({ key: 'aaa' }),
      expect.objectContaining({ key: 'aaaaa' }),
    ]);
  });

  it('renders search value when not found', () => {
    const wrapper = mount(
      <Select mode="tags" value="22" searchValue="2" open>
        <Option value="1">1</Option>
      </Select>,
    );

    expect(wrapper.find('List').props().data).toEqual([
      expect.objectContaining({
        data: expect.objectContaining({ value: '2' }),
      }),
      expect.objectContaining({
        data: expect.objectContaining({ value: '22' }),
      }),
    ]);
  });

  it('renders options matched with optionFilterProp', () => {
    const wrapper = mount(
      <Select open value="22" mode="tags" searchValue="option-1" optionFilterProp="children">
        <Option value="1">option-1</Option>
        <Option value="2">option-2</Option>
      </Select>,
    );

    expect(wrapper.find('List').props().data).toEqual([
      expect.objectContaining({
        data: expect.objectContaining({ value: '1' }),
      }),
    ]);
  });

  it('use filterOption', () => {
    const filterOption = (inputValue, option) =>
      option.value.toLowerCase().indexOf(inputValue.toLowerCase()) !== -1;

    const wrapper = mount(
      <Select mode="tags" searchValue="red" filterOption={filterOption} open>
        <Option value="Red">Red</Option>
      </Select>,
    );

    expect(wrapper.find('List').props().data).toHaveLength(2);
  });

  it('filterOption is false', () => {
    const onChange = jest.fn();
    const wrapper = mount(
      <Select mode="tags" filterOption={false} onChange={onChange}>
        <Option value="1">1</Option>
        <Option value="2">2</Option>
      </Select>,
    );

    wrapper.find('input').simulate('change', { target: { value: 'a' } });
    expect(wrapper.find('List').props().data).toHaveLength(3);

    wrapper.find('input').simulate('keyDown', { which: KeyCode.ENTER });
    expect(onChange).toHaveBeenCalledWith(['a'], expect.anything());
  });

  describe('tagRender', () => {
    it('can render custom tags', () => {
      const onTagRender = jest.fn();
      const tagRender = (props: any) => {
        const { label } = props;
        onTagRender(label);
        return (
          <span className={classNames(label, 'customize-tag')}>
            {label}
            {label}
          </span>
        );
      };
      const wrapper = mount(<Select mode="tags" tokenSeparators={[',']} tagRender={tagRender} />);

      wrapper.find('input').simulate('change', { target: { value: '1,A,42' } });
      wrapper.find('span.A').simulate('mousedown');

      expectOpen(wrapper, true);
      expect(wrapper.find('span.A').length).toBe(1);
      expect(wrapper.find('span.A').text()).toBe('AA');
      expect(onTagRender).toHaveBeenCalled();
      expect(wrapper.find('.customize-tag')).toHaveLength(3);
    });

    it('disabled', () => {
      const tagRender = jest.fn();
      mount(
        <Select
          mode="tags"
          disabled
          value={['light']}
          tagRender={tagRender}
          options={[{ value: 'light' }]}
        />,
      );

      expect(tagRender).toHaveBeenCalledWith(expect.objectContaining({ closable: false }));
    });

    it('option disabled', () => {
      const tagRender = jest.fn();
      mount(
        <Select
          mode="tags"
          disabled
          value={['light']}
          tagRender={tagRender}
          options={[{ value: 'light', disabled: true }]}
        />,
      );

      expect(tagRender).toHaveBeenCalledWith(expect.objectContaining({ closable: false }));
    });
  });

  describe('OptGroup', () => {
    const createSelect = (props) => (
      <div>
        <Select mode="tags" {...props}>
          <OptGroup key="Manager" label="Manager">
            <Option key="jack" value="jack">
              Jack
            </Option>
          </OptGroup>
          <OptGroup key="Engineer" label="Engineer">
            <Option key="Yiminghe" value="Yiminghe">
              yiminghe
            </Option>
          </OptGroup>
        </Select>
      </div>
    );

    it('renders correctly', () => {
      const wrapper = mount(createSelect({ value: ['jack', 'foo'] }));
      toggleOpen(wrapper);
      expect(wrapper.render()).toMatchSnapshot();
    });

    it('renders inputValue correctly', () => {
      const wrapper = mount(createSelect({}));
      toggleOpen(wrapper);

      wrapper.find('input').simulate('change', { target: { value: 'foo' } });
      expect(wrapper.find('List').props().data).toHaveLength(1);

      wrapper.find('input').simulate('keyDown', { which: KeyCode.ENTER });
      expect(wrapper.find('List').props().data).toHaveLength(5);
    });

    it('should work fine when filterOption function exists', () => {
      const LegacyOption = Select.Option as any; // Compatible to legacy usage

      const children = [];
      for (let i = 10; i < 36; i += 1) {
        children.push(
          <LegacyOption key={i.toString(36) + i} disabled={!(i % 3)}>
            {i.toString(36) + i}
          </LegacyOption>,
        );
      }
      const wrapper = mount(
        <Select
          mode="tags"
          style={{ width: '100%' }}
          placeholder="Tags Mode"
          filterOption={(input, { key }) => String(key).indexOf(input) >= 0}
        >
          {children}
        </Select>,
      );
      toggleOpen(wrapper);
      wrapper.find('input').simulate('change', { target: { value: 'f' } });
      expect(wrapper.find('List').props().data).toHaveLength(2);
      wrapper.find('input').simulate('keyDown', { which: KeyCode.ENTER });
      expect(findSelection(wrapper).text()).toEqual('f');
    });
  });

  it('not modify origin options', () => {
    const errSpy = jest.spyOn(console, 'error');
    const wrapper = mount(<Select mode="tags" filterOption={false} options={[]} />);
    toggleOpen(wrapper);
    wrapper.find('input').simulate('change', { target: { value: 'a' } });
    wrapper.find('input').simulate('change', { target: { value: 'ab' } });
    wrapper.find('input').simulate('change', { target: { value: 'a' } });
    wrapper.find('input').simulate('change', { target: { value: '' } });

    expect(errSpy).not.toHaveBeenCalled();
    errSpy.mockRestore();
  });

  it('correctly handles the `tabIndex` prop', () => {
    const wrapper = mount(<Select mode="tags" tabIndex={0} />);
    expect(wrapper.find('.rc-select').getDOMNode().getAttribute('tabindex')).toBeNull();

    expect(
      wrapper.find('input.rc-select-selection-search-input').getDOMNode().getAttribute('tabindex'),
    ).toBe('0');
  });

  it('press enter should not submit form', () => {
    const wrapper = mount(<Select mode="tags" open={false} />);
    const preventDefault = jest.fn();
    wrapper.find('input').simulate('keyDown', {
      which: KeyCode.ENTER,
      preventDefault,
    });

    expect(preventDefault).toHaveBeenCalled();
  });
});
