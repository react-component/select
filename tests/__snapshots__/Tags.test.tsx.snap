// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select.Tags OptGroup renders correctly 1`] = `
<div>
  <div
    class="rc-select rc-select-multiple rc-select-open rc-select-show-search"
  >
    <div
      class="rc-select-selector"
    >
      <div
        class="rc-select-selection-overflow"
      >
        <div
          class="rc-select-selection-overflow-item"
          style="opacity: 1;"
        >
          <span
            class="rc-select-selection-item"
            title="Jack"
          >
            <span
              class="rc-select-selection-item-content"
            >
              Jack
            </span>
            <span
              aria-hidden="true"
              class="rc-select-selection-item-remove"
              style="user-select: none;"
              unselectable="on"
            >
              <span
                class="rc-select-selection-item-remove-icon"
              >
                ×
              </span>
            </span>
          </span>
        </div>
        <div
          class="rc-select-selection-overflow-item"
          style="opacity: 1;"
        >
          <span
            class="rc-select-selection-item"
            title="foo"
          >
            <span
              class="rc-select-selection-item-content"
            >
              foo
            </span>
            <span
              aria-hidden="true"
              class="rc-select-selection-item-remove"
              style="user-select: none;"
              unselectable="on"
            >
              <span
                class="rc-select-selection-item-remove-icon"
              >
                ×
              </span>
            </span>
          </span>
        </div>
        <div
          class="rc-select-selection-overflow-item rc-select-selection-overflow-item-suffix"
          style="opacity: 1;"
        >
          <div
            class="rc-select-selection-search"
            style="width: 0px;"
          >
            <input
              aria-activedescendant="rc_select_TEST_OR_SSR_list_1"
              aria-autocomplete="list"
              aria-controls="rc_select_TEST_OR_SSR_list"
              aria-expanded="true"
              aria-haspopup="listbox"
              aria-owns="rc_select_TEST_OR_SSR_list"
              autocomplete="off"
              class="rc-select-selection-search-input"
              id="rc_select_TEST_OR_SSR"
              role="combobox"
              type="search"
              value=""
            />
            <span
              aria-hidden="true"
              class="rc-select-selection-search-mirror"
            >
               
            </span>
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        class="rc-select-dropdown"
        style="opacity: 0; pointer-events: none; min-width: 0; width: 0px;"
      >
        <div>
          <div
            id="rc_select_TEST_OR_SSR_list"
            role="listbox"
            style="height: 0px; width: 0px; overflow: hidden;"
          >
            <div
              aria-selected="false"
              id="rc_select_TEST_OR_SSR_list_0"
              role="option"
            />
            <div
              aria-label="Jack"
              aria-selected="true"
              id="rc_select_TEST_OR_SSR_list_1"
              role="option"
            >
              jack
            </div>
            <div
              aria-selected="false"
              id="rc_select_TEST_OR_SSR_list_2"
              role="option"
            />
          </div>
          <div
            class="rc-virtual-list"
            style="position: relative;"
          >
            <div
              class="rc-virtual-list-holder"
              style="max-height: 200px; overflow-y: auto;"
            >
              <div>
                <div
                  class="rc-virtual-list-holder-inner"
                  style="display: flex; flex-direction: column;"
                >
                  <div
                    class="rc-select-item rc-select-item-group"
                  >
                    Manager
                  </div>
                  <div
                    aria-selected="true"
                    class="rc-select-item rc-select-item-option rc-select-item-option-grouped rc-select-item-option-active rc-select-item-option-selected"
                    title="Jack"
                  >
                    <div
                      class="rc-select-item-option-content"
                    >
                      Jack
                    </div>
                    <span
                      aria-hidden="true"
                      class="rc-select-item-option-state"
                      style="user-select: none;"
                      unselectable="on"
                    >
                      <span
                        class="rc-select-item-option-state-icon"
                      >
                        ✓
                      </span>
                    </span>
                  </div>
                  <div
                    class="rc-select-item rc-select-item-group"
                  >
                    Engineer
                  </div>
                  <div
                    aria-selected="false"
                    class="rc-select-item rc-select-item-option rc-select-item-option-grouped"
                    title="yiminghe"
                  >
                    <div
                      class="rc-select-item-option-content"
                    >
                      yiminghe
                    </div>
                    <span
                      aria-hidden="true"
                      class="rc-select-item-option-state"
                      style="user-select: none;"
                      unselectable="on"
                    >
                      <span
                        class="rc-select-item-option-state-icon"
                      />
                    </span>
                  </div>
                  <div
                    aria-selected="true"
                    class="rc-select-item rc-select-item-option rc-select-item-option-selected"
                    title="foo"
                  >
                    <div
                      class="rc-select-item-option-content"
                    >
                      foo
                    </div>
                    <span
                      aria-hidden="true"
                      class="rc-select-item-option-state"
                      style="user-select: none;"
                      unselectable="on"
                    >
                      <span
                        class="rc-select-item-option-state-icon"
                      >
                        ✓
                      </span>
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Select.Tags render not display maxTagPlaceholder if maxTagCount not reach 1`] = `
<div
  class="rc-select rc-select-multiple rc-select-show-search"
>
  <div
    class="rc-select-selector"
  >
    <div
      class="rc-select-selection-overflow"
    >
      <div
        class="rc-select-selection-overflow-item rc-select-selection-overflow-item-suffix"
        style="opacity: 1;"
      >
        <div
          class="rc-select-selection-search"
          style="width: 0px;"
        >
          <input
            aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
            aria-autocomplete="list"
            aria-controls="rc_select_TEST_OR_SSR_list"
            aria-haspopup="listbox"
            aria-owns="rc_select_TEST_OR_SSR_list"
            autocomplete="off"
            class="rc-select-selection-search-input"
            id="rc_select_TEST_OR_SSR"
            role="combobox"
            type="search"
            value=""
          />
          <span
            aria-hidden="true"
            class="rc-select-selection-search-mirror"
          >
             
          </span>
        </div>
      </div>
    </div>
    <span
      class="rc-select-selection-placeholder"
    />
  </div>
</div>
`;

exports[`Select.Tags render truncates tags by maxTagCount and show maxTagPlaceholder 1`] = `
<div
  class="rc-select rc-select-multiple rc-select-show-search"
>
  <div
    class="rc-select-selector"
  >
    <div
      class="rc-select-selection-overflow"
    >
      <div
        class="rc-select-selection-overflow-item"
        style="opacity: 1;"
      >
        <span
          class="rc-select-selection-item"
          title="One"
        >
          <span
            class="rc-select-selection-item-content"
          >
            One
          </span>
          <span
            aria-hidden="true"
            class="rc-select-selection-item-remove"
            style="user-select: none;"
            unselectable="on"
          >
            <span
              class="rc-select-selection-item-remove-icon"
            >
              ×
            </span>
          </span>
        </span>
      </div>
      <div
        class="rc-select-selection-overflow-item"
        style="opacity: 1;"
      >
        <span
          class="rc-select-selection-item"
          title="Two"
        >
          <span
            class="rc-select-selection-item-content"
          >
            Two
          </span>
          <span
            aria-hidden="true"
            class="rc-select-selection-item-remove"
            style="user-select: none;"
            unselectable="on"
          >
            <span
              class="rc-select-selection-item-remove-icon"
            >
              ×
            </span>
          </span>
        </span>
      </div>
      <div
        class="rc-select-selection-overflow-item rc-select-selection-overflow-item-rest"
        style="opacity: 1;"
      >
        <span
          class="rc-select-selection-item"
        >
          <span
            class="rc-select-selection-item-content"
          >
            <span>
              Omitted
            </span>
          </span>
        </span>
      </div>
      <div
        class="rc-select-selection-overflow-item rc-select-selection-overflow-item-suffix"
        style="opacity: 1;"
      >
        <div
          class="rc-select-selection-search"
          style="width: 0px;"
        >
          <input
            aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
            aria-autocomplete="list"
            aria-controls="rc_select_TEST_OR_SSR_list"
            aria-haspopup="listbox"
            aria-owns="rc_select_TEST_OR_SSR_list"
            autocomplete="off"
            class="rc-select-selection-search-input"
            id="rc_select_TEST_OR_SSR"
            role="combobox"
            type="search"
            value=""
          />
          <span
            aria-hidden="true"
            class="rc-select-selection-search-mirror"
          >
             
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Select.Tags render truncates tags by maxTagCount and show maxTagPlaceholder function 1`] = `
<div
  class="rc-select rc-select-multiple rc-select-show-search"
>
  <div
    class="rc-select-selector"
  >
    <div
      class="rc-select-selection-overflow"
    >
      <div
        class="rc-select-selection-overflow-item"
        style="opacity: 1;"
      >
        <span
          class="rc-select-selection-item"
          title="One"
        >
          <span
            class="rc-select-selection-item-content"
          >
            One
          </span>
          <span
            aria-hidden="true"
            class="rc-select-selection-item-remove"
            style="user-select: none;"
            unselectable="on"
          >
            <span
              class="rc-select-selection-item-remove-icon"
            >
              ×
            </span>
          </span>
        </span>
      </div>
      <div
        class="rc-select-selection-overflow-item"
        style="opacity: 1;"
      >
        <span
          class="rc-select-selection-item"
          title="Two"
        >
          <span
            class="rc-select-selection-item-content"
          >
            Two
          </span>
          <span
            aria-hidden="true"
            class="rc-select-selection-item-remove"
            style="user-select: none;"
            unselectable="on"
          >
            <span
              class="rc-select-selection-item-remove-icon"
            >
              ×
            </span>
          </span>
        </span>
      </div>
      <div
        class="rc-select-selection-overflow-item rc-select-selection-overflow-item-rest"
        style="opacity: 1;"
      >
        <span
          class="rc-select-selection-item"
        >
          <span
            class="rc-select-selection-item-content"
          >
            <span>
              1 values omitted
            </span>
          </span>
        </span>
      </div>
      <div
        class="rc-select-selection-overflow-item rc-select-selection-overflow-item-suffix"
        style="opacity: 1;"
      >
        <div
          class="rc-select-selection-search"
          style="width: 0px;"
        >
          <input
            aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
            aria-autocomplete="list"
            aria-controls="rc_select_TEST_OR_SSR_list"
            aria-haspopup="listbox"
            aria-owns="rc_select_TEST_OR_SSR_list"
            autocomplete="off"
            class="rc-select-selection-search-input"
            id="rc_select_TEST_OR_SSR"
            role="combobox"
            type="search"
            value=""
          />
          <span
            aria-hidden="true"
            class="rc-select-selection-search-mirror"
          >
             
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Select.Tags render truncates values by maxTagTextLength 1`] = `
<div
  class="rc-select rc-select-multiple rc-select-show-search"
>
  <div
    class="rc-select-selector"
  >
    <div
      class="rc-select-selection-overflow"
    >
      <div
        class="rc-select-selection-overflow-item"
        style="opacity: 1;"
      >
        <span
          class="rc-select-selection-item"
          title="One"
        >
          <span
            class="rc-select-selection-item-content"
          >
            On...
          </span>
          <span
            aria-hidden="true"
            class="rc-select-selection-item-remove"
            style="user-select: none;"
            unselectable="on"
          >
            <span
              class="rc-select-selection-item-remove-icon"
            >
              ×
            </span>
          </span>
        </span>
      </div>
      <div
        class="rc-select-selection-overflow-item"
        style="opacity: 1;"
      >
        <span
          class="rc-select-selection-item"
          title="Two"
        >
          <span
            class="rc-select-selection-item-content"
          >
            Tw...
          </span>
          <span
            aria-hidden="true"
            class="rc-select-selection-item-remove"
            style="user-select: none;"
            unselectable="on"
          >
            <span
              class="rc-select-selection-item-remove-icon"
            >
              ×
            </span>
          </span>
        </span>
      </div>
      <div
        class="rc-select-selection-overflow-item rc-select-selection-overflow-item-suffix"
        style="opacity: 1;"
      >
        <div
          class="rc-select-selection-search"
          style="width: 0px;"
        >
          <input
            aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
            aria-autocomplete="list"
            aria-controls="rc_select_TEST_OR_SSR_list"
            aria-haspopup="listbox"
            aria-owns="rc_select_TEST_OR_SSR_list"
            autocomplete="off"
            class="rc-select-selection-search-input"
            id="rc_select_TEST_OR_SSR"
            role="combobox"
            type="search"
            value=""
          />
          <span
            aria-hidden="true"
            class="rc-select-selection-search-mirror"
          >
             
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
